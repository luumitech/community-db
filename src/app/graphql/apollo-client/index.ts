import {
  ApolloClient,
  ApolloLink,
  FetchResult,
  HttpLink,
  InMemoryCache,
  Observable,
  Operation,
  from,
  split,
} from '@apollo/client';
import { onError } from '@apollo/client/link/error';
import { createPersistedQueryLink } from '@apollo/client/link/persisted-queries';
import { getMainDefinition } from '@apollo/client/utilities';
import createUploadLink from 'apollo-upload-client/createUploadLink.mjs';
import { print } from 'graphql';
import { Client, ClientOptions, createClient } from 'graphql-sse';
import { typePolicies } from './type-policies';

const GRAPHQL_URL = '/api/graphql';

const cache = new InMemoryCache({
  typePolicies,
});

/** GraphQL over SSE, for handling subscription */
class SSELink extends ApolloLink {
  private client: Client;

  constructor(options: ClientOptions) {
    super();
    this.client = createClient(options);
  }

  public request(operation: Operation): Observable<FetchResult> {
    return new Observable((sink) => {
      return this.client.subscribe<FetchResult>(
        {
          ...operation,
          // query: print(operation.query)
          // When using persisted query, information is embedded in
          // extension object
          query: '',
        },
        {
          next: sink.next.bind(sink),
          complete: sink.complete.bind(sink),
          error: sink.error.bind(sink),
        }
      );
    });
  }
}

/** Unified error handling for apollo */
const errorLink = onError((response) => {
  const { graphQLErrors, networkError, operation } = response;
  if (graphQLErrors) {
    graphQLErrors.map(({ message, locations, path, extensions }) =>
      console.warn(
        `[GraphQL error]: Message: ${message}, Path: ${path}, Location: %o, extensions: %o`,
        locations,
        extensions
      )
    );
  }
  if (networkError) {
    console.warn(networkError);
  }
  console.warn('operation: %o', operation);
});

const persistedQuerylink = createPersistedQueryLink({
  // @ts-expect-error __meta__ inserted by graphql yoga
  generateHash: (document) => document.__meta__.hash,
});

// Non-subscription enabled link
// const httpLink = new HttpLink({
const httpLink = createUploadLink({ uri: GRAPHQL_URL });

// Subscription enabled link
const sseLink = new SSELink({ url: GRAPHQL_URL });

const splitLink = split(
  ({ query }) => {
    const definition = getMainDefinition(query);
    return (
      definition.kind === 'OperationDefinition' &&
      definition.operation === 'subscription'
    );
  },
  from([sseLink]),
  from([errorLink, httpLink])
);

const apolloClient = new ApolloClient({
  link: from([persistedQuerylink, splitLink]),
  cache,
});

export default apolloClient;
