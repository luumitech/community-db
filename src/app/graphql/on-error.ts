import { ApolloError } from '@apollo/client';
import { toast } from '~/view/base/toastify';

interface OnErrorOpt {
  /**
   * Handle error on your own, if the error is not handled, return it, and it
   * will be handled by the default error handler
   */
  customOnError?: (err: ApolloError) => ApolloError | undefined | void;
}

export function onError(error: ApolloError, opt?: OnErrorOpt) {
  if (opt?.customOnError == null) {
    toast.error(error.message);
    return;
  }

  // Custom error handler
  const unhandledError = opt.customOnError(error);
  if (unhandledError) {
    // Default error goes to toast
    toast.error(unhandledError.message);
  }
}
