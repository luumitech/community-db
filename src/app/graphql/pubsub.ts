/**
 * Graphql-Yoga pubsub
 */
import { createPubSub } from '@graphql-yoga/subscription';
import type { Community, Property } from '@prisma/client';

// See: https://the-guild.dev/graphql/yoga-server/tutorial/advanced/02-subscriptions#subscribing-to-new-link-elements
export enum MutationType {
  CREATED = 'CREATED',
  UPDATED = 'UPDATED',
  DELETED = 'DELETED',
}

export interface PubSubEvent {
  /**
   * email (AuthContext) of broadcaster
   */
  broadcasterId: string;
  mutationType: MutationType;
}

export interface PubSubCommunityEvent extends PubSubEvent {
  community: Community;
}

export interface PubSubPropertyEvent extends PubSubEvent {
  property: Property;
}

interface PubSubEventsImpl {
  // i.e. /community/${communityId}/
  [key: `community/${string}/`]: [PubSubCommunityEvent];
  // i.e. /community/${communityId}/property
  [key: `community/${string}/property`]: [PubSubPropertyEvent];
}
type ValueOf<T> = T[keyof T];

type PubSubEvents = PubSubEventsImpl &
  Record<string, ValueOf<PubSubEventsImpl>>;

export const pubSub = createPubSub<PubSubEvents>();
