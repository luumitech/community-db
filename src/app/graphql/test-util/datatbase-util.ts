import { Prisma } from '@prisma/client';
import * as XLSX from 'xlsx';
import { importLcraDB } from '~/lib/lcra-community/import';
import prisma from '~/lib/prisma';

/** Database related utilities for testing purpose */
export class DatabaseUtil {
  /**
   * Seed database with excel file in `fixturePath`
   *
   * @param fixturePath Excel workbook
   * @returns Database entries returned
   */
  async seed(fixturePath: string) {
    const workbook = XLSX.readFile(fixturePath);
    const importResult = importLcraDB(workbook);
    const { propertyList, ...others } = importResult;
    const communitySeed: Prisma.CommunityCreateInput[] = [
      {
        name: 'Test Community',
        ...others,
        propertyList: {
          create: propertyList,
        },
      },
    ];

    const accessSeed: Prisma.AccessCreateWithoutUserInput[] = communitySeed.map(
      (community) => ({
        role: 'ADMIN',
        community: {
          create: community,
        },
      })
    );

    await prisma.user.create({
      data: {
        email: 'jest@email.com',
        accessList: {
          create: accessSeed,
        },
      },
    });

    return importResult;
  }
}
