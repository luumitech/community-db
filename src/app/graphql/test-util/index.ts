import { GraphQLUtil } from './graphql-util';

/**
 * Helper test utility class for initializaton of key
 * test tools when running Jest unit tests
 */
export class TestUtil {
  public graphql: GraphQLUtil;

  constructor() {
    this.graphql = new GraphQLUtil();
  }

  /**
   * Initialization routine for setting services required to run jest tests
   * - clears and initialize mongodb
   * - setup blob service
   *
   * This is expected to be constructed in `beforeAll` of each test suite
   * to initialize the mongoose connection
   */
  async initialize() {}

  /**
   * Termination routine to clean up services used during jest tests
   *
   * This is expected to be constructed in `afterAll` of each test suite
   * to terminate the mongoose connection
   */
  async terminate() {}
}
