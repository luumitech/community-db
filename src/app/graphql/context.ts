import { GraphQLError } from 'graphql';
import { type YogaInitialContext } from 'graphql-yoga';
import { getServerSession } from 'next-auth/next';
import { authOptions } from '~/api/auth/[...nextauth]/auth-options';
import prisma from '../lib/prisma';
import { pubSub } from './pubsub';

export async function createContext(ctx: YogaInitialContext) {
  const session = await getServerSession(authOptions);

  // All graphQL operations require user to be authenticated
  if (!session) {
    throw new GraphQLError('You are not authenticated');
  }

  const { user } = session;
  if (!user) {
    throw new GraphQLError('Auth: Missing user context');
  }
  const { uid } = user;
  if (!uid) {
    throw new GraphQLError('Auth: Missing uid in user context');
  }
  return {
    user,
    pubSub,
  };
}

export type Context = Awaited<ReturnType<typeof createContext>>;
