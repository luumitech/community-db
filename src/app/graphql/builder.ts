import SchemaBuilder from '@pothos/core';
import PrismaPlugin from '@pothos/plugin-prisma';
import RelayPlugin from '@pothos/plugin-relay';
import { DateResolver, DateTimeResolver } from 'graphql-scalars';
import prisma from '../lib/prisma';
import { createContext } from './context';
import type PrismaTypes from './generated/pothos-types.ts';

export const builder = new SchemaBuilder<{
  PrismaTypes: PrismaTypes;
  Context: ReturnType<typeof createContext>;
  Scalars: {
    Date: { Input: Date; Output: Date };
    DateTime: { Input: Date; Output: Date };
  };
  DefaultEdgesNullability: false;
}>({
  plugins: [PrismaPlugin, RelayPlugin],
  relayOptions: {
    cursorType: 'String',
    edgesFieldOptions: {
      nullable: false,
    },
  },
  prisma: {
    client: prisma,
  },
});

builder.addScalarType('Date', DateResolver, {});
builder.addScalarType('DateTime', DateTimeResolver, {});
builder.queryType({});
builder.mutationType({});
