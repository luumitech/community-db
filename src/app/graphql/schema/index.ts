import { builder } from '../builder';

import './access';
import './community';
import './geocode';
import './offset-pagination';
import './payment';
import './property';
import './subscription';
import './user';

export const schema = builder.toSchema();
