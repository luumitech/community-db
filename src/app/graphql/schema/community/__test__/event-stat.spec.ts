import path from 'path';
import * as XLSX from 'xlsx';
import { graphql } from '~/graphql/generated';
import { TestUtil } from '~/graphql/test-util';
import { ITEM_DELIMITER } from '~/lib/lcra-community/delimiter-util';
import { toTicketList } from '~/lib/lcra-community/ticket-list-util';
import { WorksheetHelper } from '~/lib/worksheet-helper';

const EventStatQuery = graphql(/* GraphQL */ `
  query EventStatSpec_EventStat($year: Int!) {
    userCurrent {
      accessList {
        community {
          id
          communityStat {
            eventStat(year: $year) {
              eventName
              ticketList {
                ticketName
                count
                price
                paymentMethod
              }
            }
          }
        }
      }
    }
  }
`);

/** Find column index header matches the specified `val` */
function findColIdx(ws: WorksheetHelper, val: string) {
  for (const cell of ws.iterColumn('1')) {
    if (typeof cell.v === 'string' && cell.v === val) {
      return cell.colIdx;
    }
  }
}

describe('Event Statistices', () => {
  const testUtil = new TestUtil();
  const workbook = XLSX.readFile(
    path.join(process.cwd(), '__fixtures__', 'simple.xlsx')
  );

  beforeAll(async () => {
    await testUtil.initialize();
  });

  afterAll(async () => {
    await testUtil.terminate();
  });

  afterEach(async () => {
    await testUtil.database.dropDatabase();
  });

  test('Verify event statistics', async () => {
    const wb = testUtil.database.cloneWorkbook(workbook);
    await testUtil.database.seedFromWorkbook(wb);

    const result = await testUtil.graphql.executeSingle({
      document: EventStatQuery,
      variables: { year: 2023 },
    });
    const eventStat =
      result.data?.userCurrent.accessList[0].community.communityStat.eventStat;
    expect(eventStat).toBeDefined();
    expect(eventStat).toMatchSnapshot();
  });

  test('Ticket with zero count and zero price should be ignored', async () => {
    const wb = testUtil.database.cloneWorkbook(workbook);
    const ws = WorksheetHelper.fromFirstSheet(wb);
    const colIdx = findColIdx(ws, 'Y23-ticket');

    expect(colIdx).toBeDefined();
    const row2 = ws.cell(colIdx!, '2');
    const row3 = ws.cell(colIdx!, '3');
    row2.v = toTicketList([
      { ticketName: 'meal', count: 0, price: '0', paymentMethod: '' },
      { ticketName: 'drink', count: 1, price: '1', paymentMethod: 'cash' },
    ]);
    // No ticket information for row 3 (for both events)
    row3.v = [toTicketList([]), toTicketList([])].join(ITEM_DELIMITER);

    await testUtil.database.seedFromWorkbook(wb);
    const result = await testUtil.graphql.executeSingle({
      document: EventStatQuery,
      variables: { year: 2023 },
    });
    const eventStat =
      result.data?.userCurrent.accessList[0].community.communityStat.eventStat;
    expect(eventStat).toBeDefined();
    expect(eventStat).toMatchSnapshot();
  });

  test('Ticket name with delimiter characters should be filtered', async () => {
    const wb = testUtil.database.cloneWorkbook(workbook);
    const ws = WorksheetHelper.fromFirstSheet(wb);
    const colIdx = findColIdx(ws, 'Y23-ticket');

    expect(colIdx).toBeDefined();
    const row2 = ws.cell(colIdx!, '2');
    const row3 = ws.cell(colIdx!, '3');
    row2.v = toTicketList([
      {
        ticketName: 'me:/;al',
        count: 2,
        price: '0.03',
        paymentMethod: 'ca:s/h;',
      },
      {
        ticketName: 'd:r:/i;nk',
        count: 1,
        price: '1.23',
        paymentMethod: 'cas::h',
      },
    ]);
    // No ticket information for row 3 (for both events)
    row3.v = [toTicketList([]), toTicketList([])].join(ITEM_DELIMITER);

    await testUtil.database.seedFromWorkbook(wb);
    const result = await testUtil.graphql.executeSingle({
      document: EventStatQuery,
      variables: { year: 2023 },
    });
    const eventStat =
      result.data?.userCurrent.accessList[0].community.communityStat.eventStat;
    expect(eventStat).toBeDefined();
    expect(eventStat).toMatchSnapshot();
  });
});
