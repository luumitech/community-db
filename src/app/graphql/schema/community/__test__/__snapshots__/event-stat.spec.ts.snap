// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Event Statistices Ticket name with delimiter characters should be filtered 1`] = `
[
  {
    "__typename": "EventStat",
    "eventName": "Membership Carry Forward",
    "ticketList": [],
  },
  {
    "__typename": "EventStat",
    "eventName": "Summer Festival",
    "ticketList": [],
  },
  {
    "__typename": "EventStat",
    "eventName": "Corn Roast",
    "ticketList": [
      {
        "__typename": "TicketStat",
        "count": 2,
        "paymentMethod": "cash",
        "price": "0.03",
        "ticketName": "meal",
      },
      {
        "__typename": "TicketStat",
        "count": 1,
        "paymentMethod": "cash",
        "price": "1.23",
        "ticketName": "drink",
      },
    ],
  },
]
`;

exports[`Event Statistices Ticket with zero count and zero price should be ignored 1`] = `
[
  {
    "__typename": "EventStat",
    "eventName": "Membership Carry Forward",
    "ticketList": [],
  },
  {
    "__typename": "EventStat",
    "eventName": "Summer Festival",
    "ticketList": [],
  },
  {
    "__typename": "EventStat",
    "eventName": "Corn Roast",
    "ticketList": [
      {
        "__typename": "TicketStat",
        "count": 1,
        "paymentMethod": "cash",
        "price": "1",
        "ticketName": "drink",
      },
    ],
  },
]
`;

exports[`Event Statistices Verify event statistics 1`] = `
[
  {
    "__typename": "EventStat",
    "eventName": "Membership Carry Forward",
    "ticketList": [],
  },
  {
    "__typename": "EventStat",
    "eventName": "Summer Festival",
    "ticketList": [
      {
        "__typename": "TicketStat",
        "count": 20,
        "paymentMethod": "free",
        "price": "0",
        "ticketName": "meal",
      },
      {
        "__typename": "TicketStat",
        "count": 1,
        "paymentMethod": "cash",
        "price": "1",
        "ticketName": "cotton-candy",
      },
    ],
  },
  {
    "__typename": "EventStat",
    "eventName": "Corn Roast",
    "ticketList": [
      {
        "__typename": "TicketStat",
        "count": 10,
        "paymentMethod": "cash",
        "price": "5",
        "ticketName": "meal",
      },
      {
        "__typename": "TicketStat",
        "count": 20,
        "paymentMethod": "free",
        "price": "0",
        "ticketName": "meal",
      },
      {
        "__typename": "TicketStat",
        "count": 1,
        "paymentMethod": "cash",
        "price": "1",
        "ticketName": "drink",
      },
      {
        "__typename": "TicketStat",
        "count": 1,
        "paymentMethod": "e-Transfer",
        "price": "1",
        "ticketName": "drink",
      },
    ],
  },
]
`;
